rfc/src/rfcs/0006-unstable-api.md

```diff
@@ -0,0 +1,98 @@
+- **Feature Name:** Unstable APIs
+- **RFC Tracking Issue**: <https://github.com/model-checking/kani/issues/2279>
+- **RFC PR:** <https://github.com/model-checking/kani/pull/2281>
+- **Status:** Under Review
+- **Version:** 0
+
+-------------------
+
+## Summary
+
+Provide a standard option for users to enable experimental APIs and features in Kani,
+and ensure that those APIs are off by default.
+
+## User Impact
+
+The impact should be minimal, and it is very similar to the current experience when they are trying an experimental feature in Kani.
+ The unstable API will be given a unique identifier, which the user will enable either via command line or via command argument.
+
+## User Experience
+
+Users will have to invoke Kani with:
+```
+--enable-unstable --unstable-feature <feature_identifier>
+```
+in order to enable an API in the Kani library that is marked as unstable.
+
+In order to mark an API as unstable, we will add the following attribute to the APIs marked as unstable:
+
+```rust
+#[kani::unstable(feature="<IDENTIFIER>", issue="<TRACKING_ISSUE_NUMBER>", reason="<OPTIONAL_DESCRIPTION>")]
+pub fn unstable_api() {}
+```
+
+This is similar to the interface used by [the standard library](https://rustc-dev-guide.rust-lang.org/stability.html#unstable).
+
+If the user tries to run Kani without enabling an unstable feature that is reachable,
```

Comment:
Yes, I am planning to print a warning like the one we have for unsupported features. I'll make it clear here. It will be a new warning. I prefer not to mix the warnings though.

Reply:
I would prefer if we make this an error if it's possible (e.g. if it's reachable as determined by the MIR linker).

Reply:
Yeah, we can totally do that... @adpaco-aws, what do you think?

Reply:
Makes sense to me.

One case where we don't error at the moment (but for some reason, my intuition says we should) is when we run a harness annotated with `#[kani::stub(...,...)]` but we don't pass `--enable-stubbing`. But I don't have a good sense about how many "spurious errors" will be caused by emitting an error here. I think it'd be OK to start with errors and be prepared to downgrade to warnings if errors become too annoying.

Reply:
Sounds good... let me change the proposal and add the downgrade as future work. This actually makes my problem easier to solve. =)